### ISSUE TITLE
The searched architecture performs not well

### ISSUE THREAD

**Issue by linhaojia13** [lightaime/sgas#11]
I search with SGAS Cri2, whicn return the architecture:

`Genotype(normal=[('skip_connect', 0), ('gat', 1), ('conv_1x1', 1), ('conv_1x1', 2), ('sage', 2), ('semi_gcn', 3)], normal_concat=range(1, 5))`

Then I train the compact model by stacking this cell. However, this architecture perform not well:

**Finish! best_test_overall_acc 0.897893	 test_class_acc_when_best 0.840035**

This result is far away from that(93.07) reported in the paper.
How can I reproduce the result reported in the paper? Conduct more searches with different random seeds?

**Comment by guochengqian** [comment]
Hi @linhaojia13  which experiment is this? Could you provide the command line to reproduce your result?

**Comment by linhaojia13** [comment]
Thank you for your reply @guochengqian . I tried to run the experiments on ModelNet.
First, I search the architecture using the command line below:

`CUDA_VISIBLE_DEVICES=0 python train_search.py --use_history --random_seed --data ../../data`

Then I train the compact model by stacking this cell:

`CUDA_VISIBLE_DEVICES=0 python main_modelnet.py --phase train --arch myRun1 --num_cells 3 --init_channels 128 --k 9 --save myRun1_l3_c128_k9 `

The '--arch myRun1' here is the searched architecture in the first stage, which is as follow:

`Genotype(normal=[('skip_connect', 0), ('gat', 1), ('conv_1x1', 1), ('conv_1x1', 2), ('sage', 2), ('semi_gcn', 3)], normal_concat=range(1, 5))`

**Comment by guochengqian** [comment]
Hi @linhaojia13. 
It is nearly not possible that you can only get 84. One can see from the supplementary material that: we search architectures for 10times, and each time, our acc on ModelNet40 is always higher than 90. 
Before I dig into it, could you please repel some possible issues on your side at first? very appreciate. 
Here are my instructions for you: 


1. using our searched architecture to test your training code. (this is for making sure you can get a very similar (std<0.1) in modelNet experiment. (you can find our searched architecture here: https://github.com/lightaime/sgas/blob/17684fc4cd96e534cb04ee60a5659a590fcdfa60/gcn/gcn_point/genotypes.py#L33) 

2. make sure you can reproduce step 1 at first. 

3. after 2, you now begin to train your own nas. `CUDA_VISIBLE_DEVICES=0 python train_search.py --use_history --random_seed --data ../../data`. train it 10 times, get 10 architectures. then, do step 1 (now using your archi).

**Comment by linhaojia13** [comment]
Thank you for your advice @guochengqian . 
I will follow your instructions and then tell you my experiments result.

**Comment by linhaojia13** [comment]
Hi @guochengqian .
Following your advice, I trained the nas for 8 times and trained the compact models using these searched cells.  Besides, I trained the compact model by using the best cell that you searched. The OA(%) and search genotypes are shown as follow:

```
92.83 Cri2_ModelNet_Best = Genotype(normal=[('edge_conv', 0), ('mr_conv', 1), ('skip_connect', 1), ('gat', 2), ('conv_1x1', 0), ('gat', 3)], normal_concat=range(1, 5))
89.79 myRun1 = Genotype(normal=[('skip_connect', 0), ('gat', 1), ('conv_1x1', 1), ('conv_1x1', 2), ('sage', 2), ('semi_gcn', 3)], normal_concat=range(1, 5))
92.38 myRun2 = Genotype(normal=[('res_sage', 0), ('mr_conv', 1), ('gin', 0), ('skip_connect', 1), ('skip_connect', 0), ('mr_conv', 3)], normal_concat=range(1, 5))
91.98 myRun3 = Genotype(normal=[('res_sage', 0), ('skip_connect', 1), ('gin', 0), ('semi_gcn', 2), ('skip_connect', 0), ('conv_1x1', 2)], normal_concat=range(1, 5))
92.54 myRun4 = Genotype(normal=[('skip_connect', 0), ('skip_connect', 1), ('conv_1x1', 1), ('mr_conv', 2), ('skip_connect', 0), ('gin', 2)], normal_concat=range(1, 5))
92.75 myRun5 = Genotype(normal=[('sage', 0), ('semi_gcn', 1), ('skip_connect', 1), ('mr_conv', 2), ('gat', 2), ('gin', 3)], normal_concat=range(1, 5))
92.91 myRun6 = Genotype(normal=[('mr_conv', 0), ('edge_conv', 1), ('conv_1x1', 0), ('mr_conv', 2), ('mr_conv', 1), ('gin', 2)], normal_concat=range(1, 5))
92.42 myRun7 = Genotype(normal=[('gin', 0), ('mr_conv', 1), ('res_sage', 1), ('sage', 2), ('edge_conv', 1), ('gin', 2)], normal_concat=range(1, 5))
92.34 myRun8 = Genotype(normal=[('skip_connect', 0), ('gin', 1), ('gat', 1), ('conv_1x1', 2), ('gin', 1), ('edge_conv', 2)], normal_concat=range(1, 5))
```

Based on these results, here are my questions: 
1) Using the best architecture you searched 'Cri2_ModelNet_Best' can reach 92.83 OA, which has a slight gap with the reported results (93.07) in the paper. Maybe this gap could attribute to the randomness, do you think so?
2) The best cell (myRun6) I searched can reach 92.91 OA, which is relatively close to 93.07. However, the other searched cell perform not as well as 'myRun6'. I'm wondering that if there is a gap between the whole performance of my 7 searched cell (except 'myRun1' ) and your 10 times running.


Thank you very much!

**Comment by guochengqian** [comment]
@linhaojia13 This is due to the randomness nature of deep learning (cudnn, random initialization). all of your experiments are within a reasonable std.
Glad to see that you can reproduce our results using our searched architecture. 

 if you prefer reproducibility, you can set the random seed and turn off the cudnn benchmark like this:
```
    def set_seed(seed=0):
        random.seed(seed)
        np.random.seed(seed)
        torch.manual_seed(seed)
        torch.cuda.manual_seed(seed)
        torch.cuda.manual_seed_all(seed)
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False
```
Note: do not set seed for searching! (only set the seed for evaluating your searched cell)

**Comment by linhaojia13** [comment]
@guochengqian  Thank you for your patient answer. Now I have confidence in SGAS.
I have another question. It seems that you set "--unrolled" as False in all experiments. As we known, DARTS can search better-performed cells using second-order approximation. Why do you exclude 2nd order approximation for SGAS?

**Comment by guochengqian** [comment]
@linhaojia13 
Hi our goal is not to have the best performance but to bridge the gap between the performance of searching and evaluation. 
If your goal is to have better performance, you can do the trick. 
Please close the issue if your question is fixed. 
Thank for contributing the issue to the community. 
Best, 
Guocheng
